test/gtest/ucp/test_ucp_request.cc

```diff
@@ -100,3 +102,392 @@ UCS_TEST_P(test_ucp_request, test_request_query)
+    am_data_cb(void *arg, const void *header, size_t header_length, void *data,
+               size_t length, const ucp_am_recv_param_t *param)
+    {
+        std::vector<char> *rbuf = (std::vector<char>*)arg;
+
+        memcpy((*rbuf).data(), data, length);
+        return UCS_OK;
+    }
+
+    void *send_am(std::vector<char> &sbuf, std::vector<char> *rbuf_p)
+    {
+        ucp_request_param_t req_param = {0};
+        static unsigned am_id         = 1;
+        ucp_am_handler_param_t param;
+        void *sreq;
+
+        param.field_mask = UCP_AM_HANDLER_PARAM_FIELD_ID |
+                           UCP_AM_HANDLER_PARAM_FIELD_CB |
+                           UCP_AM_HANDLER_PARAM_FIELD_ARG;
+        param.id         = am_id;
+        param.cb         = am_data_cb;
+        param.arg        = rbuf_p;
+
+        ucs_status_t status;
+        status = ucp_worker_set_am_recv_handler(receiver().worker(), &param);
+        ASSERT_UCS_OK(status);
+        sreq = ucp_am_send_nbx(sender().ep(), am_id, NULL, 0, sbuf.data(),
+                               m_msg_size, &req_param);
+        am_id++;
+        return sreq;
+    }
+
+    static void flushed_cb(ucp_request_t *request)
+    {
+        test_proto_reset *self = (test_proto_reset*)request->user_data;
+        unsigned restart_count = 0;
+        uct_pending_req_t *uct_req;
+        ucp_request_t *ucp_req;
+
+        ucs_queue_for_each_extract(uct_req, &self->m_pending, priv, 1) {
+            ucp_req = ucs_container_of(uct_req, ucp_request_t, send.uct);
+            if (ucp_req->send.state.dt_iter.offset > 0) {
+                ucp_request_send(ucp_req);
+            } else {
+                ucp_proto_request_restart(ucp_req);
+                restart_count++;
+            }
+        }
+
+        EXPECT_GT(restart_count, 0);
+        self->m_completed = 1;
+        ucp_request_complete_send(request, request->status);
+    }
+
+    void restart(ucp_ep_h ep)
+    {
+        ucp_request_param_t param;
+        param.op_attr_mask = UCP_OP_ATTR_FIELD_USER_DATA |
+                             UCP_OP_ATTR_FIELD_CALLBACK,
+        param.user_data    = this;
+        param.cb.send      = (ucp_send_nbx_callback_t)ucs_empty_function;
+
+        ucs_queue_head_init(&m_pending);
+        ucp_ep_purge_lanes(ep, ucp_request_purge_enqueue_cb, &m_pending);
+
+        void *request = ucp_ep_flush_internal(ep, 0, &param, NULL, flushed_cb,
+                                              "ep_restart");
+
+        ASSERT_FALSE(UCS_PTR_IS_ERR(request));
+        wait_for_value(&m_completed, (unsigned)1);
+        ucp_request_release(request);
+    }
+
+    void send_nb(std::vector<char> &sbuf, std::vector<char> *rbuf_p,
+                 operation_e op, bool sync, std::vector<void*> &reqs)
+    {
+        ucp_request_param_t param = {0};
+        void *rreq                = NULL;
+        void *sreq                = NULL;
+        std::vector<char> &rbuf   = *rbuf_p;
+        sbuf.resize(m_msg_size);
```

Comment:
maybe here we can just `reserve`, not `resize`?

Reply:
code was changed (now init the array inside a for loop).